### PR TITLE
Handle slot default content preview mode

### LINKED ISSUE #221: Add default content preview for plug-js

### PR DESCRIPTION
**Pull request by renan628**
## Summary
This PR adapts the preview widget to handle the slot default content preview mode. If the changes the preview mode and the locale are obtained from the URL parameters in the plugin, and forwarded to the widget. Additionally, the `index.html` used to test the widget was updated to reflect the plugin's behavior, nearing the test from the real scenario.

Related to:
- Closes #221 

### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: src/preview.ts [hunk 1]
```diff
@@ -20,6 +27,8 @@ export class PreviewPlugin implements Plugin {
         variantName: 'variant',
     };
 
+    public static readonly NONE_AUDIENCE = 'None';
```
Comment (denis-rossati) [inline src/preview.ts:30]: I guess internally we can call "NO_AUDIENCE", I'm not very sure though

```suggestion
    public static readonly NO_AUDIENCE = 'None';
```

File: preview/test/widget.spec.ts [hunk 2]
```diff
@@ -52,6 +52,23 @@ test.describe('Preview widget', () => {
         await expect(page).toHaveScreenshot('widget-expanded.png');
     });
 
+    test('should not display the experience if not specified', async ({page}) => {
+        await open(page, {
+            audience: 'A very very very very long audience name',
```
Comment (denis-rossati) [inline preview/test/widget.spec.ts:57]: I guess this should follow the 'Unknown' example, since the user will see "None" instead of a long audience

```suggestion
            audience: 'None',
```

File: preview/src/index.html [hunk 3]
```diff
@@ -18,6 +18,18 @@
 <iframe id="widget" src="./widget.html"></iframe>
 
 <script type="text/javascript">
+
```
Comment (denis-rossati) [inline preview/src/index.html:21]:
```suggestion
```

Comment (denis-rossati) [reply preview/src/index.html:21]: Also, rearrange the functions to be called only after they're declared for better readability, like this:

```
func foo() {}

func bar() {
    foo();
}
```

instead of 

```
func bar() {
    foo();
}

func foo() {}
```

File: preview/src/index.html [hunk 4]
```diff
@@ -36,23 +48,24 @@
         const metadata = payload.metadata;
 
         return {
+            audience: getAudience(metadata?.previewMode, metadata?.audienceName),
```
Comment (denis-rossati) [inline preview/src/index.html:51]:
```suggestion
        const metadata = payload.metadata;
        const audience = metadata?.previewMode === 'slotDefaultContent' ? 'None' : metadata?.audienceName ?? 'Unknown';

        return {
            audience: audience,
```

Comment (denis-rossati) [reply preview/src/index.html:51]: Actually I have divided thoughts about this suggestion, even though we can remove that function, this ternary is not very readable, which make me think that the function approach is better. I'll leave the suggestion here for you to decide

File: src/preview.ts (hunk 1 above)
Comment (renan628) [reply src/preview.ts:30]: There is a difference between "no audience" and the "none audience". "No audience" is when no audience is informed. "None audience" is when I inform I want to see the content out of any audience. Is there a slight but relevant difference. Did you get it?

File: preview/test/widget.spec.ts (hunk 2 above)
Comment (renan628) [reply preview/test/widget.spec.ts:57]: I would prefer to keep it following the other tests. I understand your idea, the experience won't be shown only in the "none audience" case. But the widget doesn't know the audience's name, it just shows what was received in the params.

File: preview/src/index.html (hunk 3 above)
Comment (renan628) [reply preview/src/index.html:21]: Some guide lines recommend doing like that, including clean code. Take a look at the third item [here](https://medium.com/oril/10-tips-to-keep-your-code-more-clean-2fa9aafea1cf#:~:text=Methods%20should%20be%20declared%20in,minor%20methods%20at%20the%20end.). Also I remember a previous PR where I was asked to do the method ordination like that :sweat_smile:.

File: preview/src/index.html (hunk 4 above)
Comment (renan628) [reply preview/src/index.html:51]: I decided by function for readability, but thanks for the suggestion.

File: preview/src/index.html (hunk 3 above)
Comment (Fryuni) [reply preview/src/index.html:21]: There is a linter for the order Denis recommended. It is even a recommended linter, and for good reason, something might call the function in between the declarations:

```ts
func bar() {
    foo();
}

bar(); // This will fail because `foo` is undefined at this point

func foo() {}
```

But this requirement is a quirk of JS and not really a code style guideline, so we explicitly disabled that linter. The recommendation is to leave the higher lever code higher in the file, so you scroll down to see the implementation detail.

File: src/preview.ts (hunk 1 above)
Comment (denis-rossati) [reply src/preview.ts:30]: Yes, thank you!

**Review by denis-rossati — APPROVED** [review]

**Review by bergaminifrg — DISMISSED** [review]

**Review by Fryuni — APPROVED** [review]

**Review by georgekaran — DISMISSED** [review]

File: preview/src/index.html [hunk 5]
```diff
@@ -36,23 +48,24 @@
...
-            audience: metadata?.audienceName ?? 'Unknown',
             variant: metadata?.variantName ?? null,
             locale: metadata?.locale ?? null,
         };
     }
 
-    function getPreviewInfo() {
-        const params = new URLSearchParams(window.location.search);
-        const token = params.get('croct-preview');
+    function getAudience(previewMode, audienceName) {
```
Comment (marcospassos) [inline preview/src/index.html:59]:
```suggestion
    function getAudienceName(previewMode, audienceName) {
```

File: src/preview.ts [hunk 6]
```diff
@@ -107,6 +116,10 @@ export class PreviewPlugin implements Plugin {
             return params;
         }
 
+        if (metadata.previewMode === ExperiencePreviewScope.SLOT_DEFAULT_CONTENT) {
+            metadata.audienceName = PreviewPlugin.NONE_AUDIENCE;
+        }
```
Comment (marcospassos) [inline src/preview.ts:121]: It should be handled by the preview widget, not by the plugin

File: src/preview.ts [hunk 7]
```diff
@@ -7,6 +7,13 @@ import {PREVIEW_WIDGET_ORIGIN, PREVIEW_WIDGET_URL} from './constants';
 const PREVIEW_PARAMETER = 'croct-preview';
 const PREVIEW_EXIT = 'exit';
 
+export enum ExperiencePreviewScope {
```
Comment (marcospassos) [inline src/preview.ts:10]: Please undo the changes in the preview plugin.

**Review by marcospassos — CHANGES_REQUESTED** [review]

File: preview/src/widget.js [hunk 8]
```diff
@@ -90,44 +90,100 @@ window.addEventListener('DOMContentLoaded', () => {
...
-        }
+    const params = new URLSearchParams(window.location.search);
 
-        document.getElementById('preview-content').textContent = params.get('variant');
-    } else {
-        document.getElementById('preview-experiment')
-            .closest('li')
-            .remove();
-    }
+    const [previewMode, experience, audience, experiment, variant, locale] = [
```
Comment (marcospassos) [inline preview/src/widget.js:171]: Why are ou using destructuring for this?

File: preview/src/widget.js [hunk 9]
```diff
@@ -90,44 +90,100 @@ window.addEventListener('DOMContentLoaded', () => {
...
+
+        if (audience === null) {
+            return;
+        }
+
+        document.getElementById('preview-audience').textContent = audience;
+    }
+
+    function handleExperiment(previewMode, variant, experiment) {
+        if (previewMode === 'slotDefaultContent' || variant === null) {
```
Comment (marcospassos) [inline preview/src/widget.js:124]: Why are you looking at the variant?

File: preview/src/widget.js [hunk 10]
```diff
@@ -90,44 +90,100 @@ window.addEventListener('DOMContentLoaded', () => {
...
+    const [previewMode, experience, audience, experiment, variant, locale] = [
+        params.get('previewMode'),
+        params.get('experience'),
+        params.get('audience'),
+        params.get('experiment'),
+        params.get('variant'),
+        params.get('locale'),
+    ]
+
+    handleExperience(previewMode, experience);
```
Comment (marcospassos) [inline preview/src/widget.js:180]: `renderAudience`, `renderExperiment`, etc...

File: preview/src/widget.js [hunk 11]
```diff
@@ -90,44 +90,100 @@ window.addEventListener('DOMContentLoaded', () => {
...
+        if(previewMode === 'slotDefaultContent') {
+            document.getElementById('preview-experience')
+                .closest('li')
+                .remove();
+
+            return;
+        }
+
+        if (experience === null) {
+            return;
```
Comment (marcospassos) [inline preview/src/widget.js:103]: No reason for early return if you have nothing after this

File: preview/src/widget.js [hunk 12]
```diff
@@ -90,44 +90,100 @@ window.addEventListener('DOMContentLoaded', () => {
...
+
+    function handleAudience(previewMode, audience) {
+        if(previewMode === 'slotDefaultContent') {
+            document.getElementById('preview-audience').textContent = 'None';
+
+            return;
+        }
+
+        if (audience === null) {
+            return;
```
Comment (marcospassos) [inline preview/src/widget.js:117]: same

File: preview/src/widget.js [hunk 13]
```diff
@@ -90,44 +90,100 @@ window.addEventListener('DOMContentLoaded', () => {
...
+                .closest('li')
+                .remove();
+
+            return;
+        }
 
-    if (params.has('experience')) {
-        document.getElementById('preview-experience').textContent = params.get('experience');
+        if (experiment === null) {
+            return;
```
Comment (marcospassos) [inline preview/src/widget.js:133]: same

**Review by marcospassos — CHANGES_REQUESTED** [review]

File: preview/src/widget.js (hunk 9 above)
Comment (renan628) [reply preview/src/widget.js:124]: I just kept this behavior, it was a [nested IF](https://github.com/croct-tech/plug-js/blob/master/preview/src/widget.js#L122). If no variant is received the experiment is not displayed.

File: preview/test/widget.spec.ts [hunk 14]
```diff
@@ -52,11 +52,14 @@ test.describe('Preview widget', () => {
         await expect(page).toHaveScreenshot('widget-expanded.png');
     });
 
-    test('should not display the experiment if not specified', async ({page}) => {
+    test('should display the none audience for the slot default content preview', async ({page}) => {
```
[the remark this answers is not in the document]
Comment (renan628) [reply preview/test/widget.spec.ts:55]: 'Not display' is being used to express that nothing will be displayed, it is not the case here because we display the 'None' label in the audience (content out of any audience). What about:
`should not specify the audience when previewing the default content`?

[the remark this answers is not in the document]
Comment (marcospassos) [reply preview/test/widget.spec.ts:55]: How about this?

> should not indicate any specific audience when previewing the default content

File: preview/src/widget.js (hunk 9 above)
Comment (marcospassos) [reply preview/src/widget.js:116]: I think it'd make more sense to handle experiment and variant together, in the same function

Comment (Fryuni) [reply preview/src/widget.js:116]: It is possible to have a variant and not an experiment name on the token.

You can create a preview for an experience draft with an experience without having set the name of the experiment.

The draft cannot be published yet, but you can get a preview of it, and there is no name to add to it.

File: preview/src/widget.js [hunk 15]
```diff
@@ -90,44 +90,77 @@ window.addEventListener('DOMContentLoaded', () => {
...
+        if(previewMode === 'slotDefaultContent') {
+            document.getElementById('preview-audience').textContent = 'None';
+
+            return;
+        }
+
+        document.getElementById('preview-audience').textContent = audience;
+    }
+
+    function renderContentAndExperiment(previewMode, variant, experiment) {
```
Comment (marcospassos) [inline preview/src/widget.js:115]:
```suggestion
    function renderExperiment(previewMode, variant, experiment) {
```

File: preview/test/widget.spec.ts-snapshots/widget-without-locale-webkit.png
Comment (marcospassos) [inline preview/test/widget.spec.ts-snapshots/widget-without-locale-webkit.png:1]: You removed the screenshot tests for the truncation behavior. We should have one screenshot test for each possible case (with locale, without locale, with experiment, etc) plus one for testing the truncation of all fields.

**Review by marcospassos — CHANGES_REQUESTED** [review]

File: preview/src/widget.js (hunk 15 above)
Comment (marcospassos) [reply preview/src/widget.js:115]: Keep them separed

Comment (marcospassos) [reply preview/src/widget.js:115]: I'm not sure if dividing the code into different functions led to any improvement

**Review by marcospassos — APPROVED** [review]
